README

Authen::Krb5::Admin is an object-oriented interface to the Kerberos 5
admin server.  Currently only MIT KDCs are supported, but the author
envisions seamless integration with other KDCs.


PREREQUISITES

Authen::Krb5::Admin requires Perl 5.005 or greater and Jeff Horwitz's
Authen::Krb5 module from CPAN.


INSTALLATION

To install Authen::Krb5::Admin without testing, run the standard
sequence:

	perl Makefile.PL
	make
	make install

The first step will prompt you for a principal with administrative
privileges on the KDC for running the test suite.  Re-run this step to
change which principal to use.

The build script will attempt to use the krb5-config tool to discover
the location of your Kerberos installation. This tool is part of the
krb5-devel package on Redhat and Fedora systems and is in the
libkrb5-dev package on Debian and Ubuntu systems. If it is not located
in /usr/local/bin/krb5-config or /usr/bin/krb5-config you can specify
the location using the KRB5_CONFIG environment variable.

If you have a non-standard installation you can override the following
locations by setting environment variables before running the
Makefile.PL script. The PREFIX path is found by using the krb5-config
tool.

   * KRB5_BINDIR - PREFIX/bin - Kerberos binaries
   * KRB5_LIBDIR - PREFIX/lib - Kerberos libraries
   * KRB5_INCDIR - PREFIX/include - Kerberos headers

Currently the install scripts do NOT correctly build for x86_64
systems where the libraries are installed in /usr/lib64 (e.g. Redhat
and Fedora). The workaround is to set the KRB5_LIBDIR environment
variable before running the Makefile.PL script, e.g.

export KRB5_LIBDIR=/usr/lib64

This will be corrected in a subsequent release.

TESTING

To run the test suite as part of the install, use

	perl Makefile.PL
	make test
	make install

This step requires administrative access to the KDC by the principal
specified in the first step.  It will create a principal, by default
"perl_test", which should NOT already exist on the KDC.  If you
already have a principal called perl_test, change Makefile.PL to use a
different test principal, and start over.


Stephen Quinney <squinney@inf.ed.ac.uk>
University of Edinburgh School of Informatics

Andrew J. Korty <ajk@iu.edu>
Indiana University


ChangeLog
VERSION 0.16
============

2012-12-13  Stephen Quinney <squinney@inf.ed.ac.uk>

	* Makefile.PL: Renamed the KRB5_CONFIG environment variable to
	   KRB5_CONFTOOL to avoid potential for confusion. Thanks to Tom
	   Poage for the bug report, this closes: #81146.

VERSION 0.15
============

2012-11-01  Stephen Quinney <squinney@inf.ed.ac.uk>

	* Added a db_args method to Authen::Krb5::Admin::Principal, thanks
	   to Dorian Taylor for the patch. This closes: #80205. This is
	   only available with MIT Kerberos version 1.7 and newer.

VERSION 0.14 (not released)
============

2012-08-06  Stephen Quinney <squinney@inf.ed.ac.uk>

	* Implemented support for the new administrative lockout options
           ( pw_max_fail, pw_failcnt_interval and pw_lockout_duration)
	   see http://k5wiki.kerberos.org/wiki/Projects/Lockout
	   for full details. This closes #76830.

VERSION 0.13
============

2012-04-26  Stephen Quinney <squinney@inf.ed.ac.uk>

	* Use the Kerberos admin.h rather than a local copy where
	   possible. Closes: #49964
	* Handle the API change which requires the context argument be
	   passed to the various init functions. There may be a better way
	   to do this but it requires a change to Authen::Krb5.
	* Reduced the set of libraries we link against as we do not use
  	   libkrb5 and libk5crypto directly, patch taken from Debian.
	* Applied patch from Kevan Carstensen to export the
	   KRB5_KDB_ACCESS_ERROR constant. Closes: #77093.

VERSION 0.12
============

2012-03-12  Stephen Quinney <squinney@inf.ed.ac.uk>

	* New maintainer, thanks to Andrew Korty for all his good work.
	* No code changes in this release.
	* Closes: #69831

VERSION 0.11
============

2008-02-25  Andrew J. Korty  <ajk@iu.edu>

	* Makefile.PL: Exit 0 if Kerberos 5 can't be found so as not to
	produce a FAIL from CPAN Testers.

VERSION 0.10
============

2008-01-30  Andrew J. Korty  <ajk@iu.edu>

	* MANIFEST, MANIFEST.SKIP, META.yml, Makefile.PL: Include and
	verify a module signature.

	* t/34-cpw.t: Skip tests that tickle what might be an
	incompatibility between Authen::Krb5 1.7 and MIT 1.6.3 that causes
	get_in_tkt_with_password to segfault.

	* Makefile.PL: Using $KRB5_EXTRAINCS, include files from
	$PREFIX/include/et.  On RHEL 5, this directory is where com_err.h
	is kept.

VERSION 0.09
============

2007-01-05  Andrew Korty  <ajk@iu.edu>

	* ppport.h: new version, complete with INT2PTR()
	* Admin.xs, typemap: Convert to four-column indent.
	* Admin.xs: Copy the policy name we get from kadm5_get_principal()
	into memory allocated with New() and then free() what was given to
	us.  That way, we can just Safefree() it in DESTROY().

2006-12-28  Andrew J. Korty  <ajk@iu.edu>

	* Admin.xs, typemap: Use INT2PTR() and friends to make sure void *
	and Perl values are compatible when perl is built with
	-Duse64bitint.

	* Makefile.PL, t/01-config.t, t/10-getprivs.t, t/20-addpol.t,
	t/21-getpol.t, t/22-getpols.t, t/23-modpol.t, t/30-addprinc.t,
	t/31-getprinc.t, t/32-getprincs.t, t/33-modprinc.t, t/34-cpw.t,
	t/35-randkey.t, t/36-rename.t, t/91-delprinc.t, t/92-delpol.t: Let
	us specify a credentials cache with the $PERL_KADM5_TEST_CACHE
	environment variable.

	* t/91-delprinc.t, t/92-delpol.t: Test to see if the
	principal/policy is gone after we delete it.

	* Admin.pm, Admin.xs, admin.h: Attempt to detect MIT Kerberos 5
	Release 1.5 and later, whose kadm5_init_with_* functions take an
	extra argument (char **db_args) and whose kadm5_config_params
	structure doesn't have a profile field.

VERSION 0.07
============

2005-02-19  Andrew J. Korty  <ajk@iu.edu>

	* Makefile.PL: Red Hat installs Kerberos in /usr/kerberos/, so
	look there too.

	* Admin.xs: Initialize structs correctly.  Calling Newz() is wrong
	because it sets all bits to zero, which doesn't guarantee that any
	element but those of type int are properly initialized.

	* Admin.pm, Admin.xs, typemap: Authen::Krb5 now has a keyblock
	class, so use that instead of a fake one.

2004-02-18  Andrew J. Korty  <ajk@iu.edu>

	* Admin.pm: Add an EXAMPLES section pointing out that examples can
	be found in the unit tests.  Add a FILES section referring to
	krb5.conf.  Add a BUGS section explaining missing functionality
	that probably belongs in Authen::Krb5.  Add perlvar(1) to the SEE
	ALSO section since it's referenced elsewhere.

VERSION 0.06
============

2004-02-12  Andrew Korty  <ajk@iu.edu>

	* Admin.pm: Document rename_principal method.

2003-07-21  Andrew Korty  <ajk@iu.edu>

	* Makefile.PL: Require Admin::Krb5 1.2.

2003-02-05  Andrew Korty  <ajk@iu.edu>

	* Makefile.PL: Add another test principal name for
	krb5_rename_principal().

	* Admin.xs: Add interface to krb5_rename_principal().

	* t/36-rename.t: Tests for rename_principal.

2003-01-30  Andrew Korty  <ajk@iu.edu>

	* Admin.pm: Principal and policy methods return TRUE or undef, not
	error codes.

	* Admin.pm, Admin.xs, admin.h: Add error table constants.

2002-10-09  Andrew Korty  <ajk@iu.edu>

	* Admin.xs: Overwrite key contents before reallocating memory.

	* Admin.pm, Admin.xs, COPYING, Makefile.PL, admin.h,
	t/00-loadme.t, t/01-config.t, t/02-policy.t, t/03-principal.t,
	t/04-key.t, t/10-getprivs.t, t/20-addpol.t, t/21-getpol.t,
	t/22-getpols.t, t/23-modpol.t, t/30-addprinc.t, t/31-getprinc.t,
	t/32-getprincs.t, t/33-modprinc.t, t/34-cpw.t, t/35-randkey.t,
	t/91-delprinc.t, t/92-delpol.t: Copyright notices.

	* Admin.pm, Admin.xs: Add documentation for recent changes.

2002-10-08  Andrew Korty  <ajk@iu.edu>

	* Admin.xs, typemap: Use Kerberos 5 types.

2002-10-04  Andrew Korty  <ajk@iu.edu>

	* Admin.xs: Do a better job of mortalizing stack values and
	freeing data.

	* Admin.xs, admin.h, t/04-key.t, t/35-randkey.t, typemap,
	Admin.pm: Add support for krb5_key_data in the form of an
	Authen::Krb5::Admin::Key class.

2002-10-03  Andrew Korty  <ajk@iu.edu>

	* t/33-modprinc.t, t/34-cpw.t: Allow authentication to our test
	principal temporarily so we can change passwords.


Makefile.PL
# Copyright (c) 2002 Andrew J. Korty
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
# SUCH DAMAGE.

# $Id: Makefile.PL,v 1.16 2008/02/25 13:43:46 ajk Exp $

use strict;

use ExtUtils::MakeMaker;

# Use the krb5-config tool if available

my $KRB5_CONFIG;
print 'checking for krb5-config ... ';
for my $path ( $ENV{KRB5_CONFIG}, '/usr/local/bin/krb5-config', '/usr/bin/krb5-config' ) {
  if ( $path && -x $path ) {
    $KRB5_CONFIG = $path;
    last;
  }
}
if ($KRB5_CONFIG) {
  print "found\n";
} else {
  print "not found, will use fallback methods\n"
}

# location of Kerberos 5 libraries and header files

print 'checking for Kerberos 5 prefix ... ';
my $PREFIX;
if ($KRB5_CONFIG) {
  chomp( $PREFIX = `$KRB5_CONFIG --prefix` );
} else {
  ($PREFIX) = grep { -r "$_/include/krb5.h" }
              qw(/opt/krb5 /usr/local /usr /usr/kerberos);
}

if ($PREFIX) {
    print "$PREFIX\n";
} else {
    print "Cannot find Kerberos 5\n";
    exit 0;
}

my $KRB5_BINDIR = $ENV{KRB5_BINDIR} || "$PREFIX/bin"    ;
my $KRB5_LIBDIR = $ENV{KRB5_LIBDIR} || "$PREFIX/lib"    ;
my $KRB5_INCDIR = $ENV{KRB5_INCDIR} || "$PREFIX/include";

# any extra libraries or include flags

my $KRB5_EXTRALIBS = '';
my $KRB5_EXTRAINCS = "-I$PREFIX/include/et";

# principal used by test suite to authenticate to kadmin server (will
# prompt if undef--you must do your own kinit if you set this)

my $KADM5_PRINCIPAL   = $ENV{PERL_KADM5_PRINCIPAL};
my $default_admin     = "$ENV{USER}/admin";

# principal/policy name used by test suite--will be created, modified,
# deleted, etc.

my $KADM5_TEST_NAME     = $ENV{PERL_KADM5_TEST_NAME  };
my $KADM5_TEST_NAME_2   = $ENV{PERL_KADM5_TEST_NAME_2};
my $default_test_name   = 'perl_test'  ;
my $default_test_name_2 = 'perl_test_2';

##### DO NOT CHANGE ANYTHING BELOW HERE #####

# check for libk5crypto -- only in krb5-1.1 and above

print 'checking for libk5crypto ... ';
my $cryptolib;
if ( -r "$KRB5_LIBDIR/libk5crypto.a" ||
     -r "$KRB5_LIBDIR/libk5crypto.so" ) {
    print "found\n";
    $cryptolib = '-lk5crypto';
} else {
    print "not found (using libcrypto)\n";
    $cryptolib = '-lcrypto';
}

print 'checking for kadm5/admin.h ... ';
my $KRB5_DEFS = '';
if ( -f "$KRB5_INCDIR/kadm5/admin.h" ) {
  print "found\n";
} else {
  print "not found, assuming this is Kerberos 1.6\n";
  $KRB5_DEFS .= ' -DUSE_LOCAL_ADMINH';
}

print 'checking for kdb.h ... ';
if ( -f "$KRB5_INCDIR/kdb.h" ) {
  print "found\n";
  $KRB5_DEFS .= ' -DHAVE_KDB_H';
} else {
  print "not found, assuming this is Kerberos 1.6\n";
}

# get values for test suite

unless ($KADM5_PRINCIPAL) {
    print 'Authenticate as user/instance[@REALM] for testing ',
          "[$default_admin]: ";
    $KADM5_PRINCIPAL = <STDIN>;
    chomp $KADM5_PRINCIPAL;
    $KADM5_PRINCIPAL ||= $default_admin;
}

unless ($KADM5_TEST_NAME) {
    print 'Use user/instance (should NOT exist!) for testing ',
          "[$default_test_name]: ";
    $KADM5_TEST_NAME = <STDIN>;
    chomp $KADM5_TEST_NAME;
    $KADM5_TEST_NAME ||= $default_test_name;
}

unless ($KADM5_TEST_NAME_2) {
    print 'Use user/instance (should NOT exist!) for testing renames ',
          "[$default_test_name_2]: ";
    $KADM5_TEST_NAME_2 = <STDIN>;
    chomp $KADM5_TEST_NAME_2;
    $KADM5_TEST_NAME_2 ||= $default_test_name_2;
}

sub MY::insert_kinit_maybe {
    my ($commands) = @_;
    my $test_cache = $ENV{PERL_KADM5_TEST_CACHE} || '/tmp/perl_test';
    my $new_vars = join ' ', "PERL_KADM5_PRINCIPAL=$KADM5_PRINCIPAL",
                             "PERL_KADM5_TEST_NAME=$KADM5_TEST_NAME",
                             "PERL_KADM5_TEST_NAME_2=$KADM5_TEST_NAME_2",
                             "PERL_KADM5_TEST_CACHE=$test_cache ";
    $commands =~ s/PERL_DL_NONLAZY=1 /$new_vars /;
    return exists $ENV{PERL_KADM5_TEST_CACHE}
         ? $commands
         : "\t$KRB5_BINDIR/kinit -S kadmin/admin -c $test_cache -l 50m "
           . "$KADM5_PRINCIPAL\n$commands"
         ;
}

sub MY::test_via_harness {
    package MY;
    return insert_kinit_maybe shift->SUPER::test_via_harness(@_);
}

sub MY::test_via_script {
    package MY;
    return insert_kinit_maybe shift->SUPER::test_via_script(@_);
}

WriteMakefile(
    DEFINE       => "$KRB5_DEFS",
    INC          => "-I$KRB5_INCDIR $KRB5_EXTRAINCS",
    LIBS         => ["-L$KRB5_LIBDIR -lkadm5clnt -lcom_err",
                     $KRB5_EXTRALIBS],
    NAME         => 'Authen::Krb5::Admin',
    NO_META      => 1,
    PREREQ_PM    => {'Authen::Krb5' => 1.2},
    VERSION_FROM => 'Admin.pm',
);
